primitives/runtime/src/offchain/storage_lock.rs

```diff
@@ -0,0 +1,468 @@
+where
+	B: BlockNumberProvider,
+	<B as BlockNumberProvider>::BlockNumber: Copy,
+{
+	/// The block number that has to be reached in order
+	/// for the lock to be considered stale.
+	pub block_number: <B as BlockNumberProvider>::BlockNumber,
+	/// Additional timestamp based deadline which has to be
+	/// reached in order for the lock to become stale.
+	pub timestamp: Timestamp,
+}
+
+// derive not possible, since `B` does not necessarily implement `trait Clone`
+impl<B> Clone for BlockAndTime<B>
+where
+	B: BlockNumberProvider,
+	<B as BlockNumberProvider>::BlockNumber: Clone,
+{
+	fn clone(&self) -> Self {
+		Self {
+			block_number: self.block_number,
+			timestamp: self.timestamp,
+		}
+	}
+}
+
+// derive not possible, since `B` does not necessarily implement `trait Copy`
+impl<B: BlockNumberProvider> Copy for BlockAndTime<B>
+{
+}
+
+impl<B: BlockNumberProvider> Lockable for BlockAndTime<B>
+{
+	fn current() -> Self {
+		Self {
+			block_number: B::current_block_number(),
+			timestamp: offchain::timestamp(),
+		}
+	}
+
+	fn deadline() -> Self {
+		Self {
+			block_number: B::deadline_block_number(),
+			timestamp: offchain::timestamp().add(Duration::from_millis(
+				STORAGE_LOCK_DEFAULT_EXPIRY_DURATION_MS,
+			)),
+		}
+	}
+
+	fn expired(&self, reference: &Self) -> bool {
+		let current = <Self as Lockable>::current();
+		current.timestamp > reference.timestamp && current.block_number > reference.block_number
+	}
+
+	fn snooze(&self, deadline: &Self) {
+		let remainder: Duration = self.timestamp.diff(&(deadline.timestamp));
+		let snooze = core::cmp::min(remainder.millis(), STORAGE_LOCK_PER_CHECK_ITERATION_SNOOZE);
+		sp_io::offchain::sleep_until(self.timestamp.add(Duration::from_millis(snooze)));
+	}
+}
+
+/// Storage based lock.
+///
+/// A lock that is persisted in the DB and provides a mutex behavior
+/// with a defined safety expirey deadline based on a [`Lockable`](Self::Lockable)
+/// implementation.
+pub struct StorageLock<'a, L>
+where
+	L: Lockable,
+{
+	// A storage value ref which defines the DB entry representing the lock.
+	value_ref: StorageValueRef<'a>,
+	deadline: L,
+}
+
+impl<'a, L> StorageLock<'a, L>
+where
+	L: Lockable,
+{
+	/// Create a new storage lock with [default expiry duration](Self::STORAGE_LOCK_DEFAULT_EXPIRY_DURATION_MS).
+	pub fn new(key: &'a [u8]) -> Self {
+		Self {
+			value_ref: StorageValueRef::<'a>::persistent(key),
+			deadline: L::deadline(),
+		}
+	}
+
+	fn try_lock_inner(&mut self, new_deadline: L) -> Result<(), Option<L>> {
+		let now = L::current();
+		let res = self
+			.value_ref
+			.mutate(|s: Option<Option<L>>| -> Result<L, Option<L>> {
+				match s {
+					// no lock set, we can safely acquire it
+					None => Ok(new_deadline),
+					// lock is set, but it's old. We can re-acquire it.
+					Some(Some(deadline)) if now.expired(&deadline) => Ok(new_deadline),
+					// lock is present and is still active
+					Some(Some(deadline)) => Err(Some(deadline)),
+					_ => Err(None),
```

Comment:
`Some(None)` case should just be the same as `None`. It means that the lock is set, but we can't read the content - imho we should simply overwrite it.